## Машинное обучение 1, ПМИ ФКН ВШЭ

## Практическое домашнее задание 1

### Общая информация

Дата выдачи: 11.09.2021

Мягкий дедлайн: 23:59MSK 25.09.2021

Жесткий дедлайн: 23:59MSK 28.09.2021

### О задании

Задание состоит из двух разделов, посвященных работе с табличными данными с помощью библиотеки pandas и визуализации с помощью matplotlib. В первом разделе вам предстоит выполнить базовые задания с помощью вышеуказанных библиотек, а во втором распределить студентов по курсам. Баллы даются за выполнение отдельных пунктов. Задачи в рамках одного раздела рекомендуется решать в том порядке, в котором они даны в задании.

Задание направлено на освоение jupyter notebook (будет использоваться в дальнейших заданиях), библиотек pandas и matplotlib.

### Оценивание и штрафы
Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 10 баллов.

Сдавать задание после жёсткого дедлайна нельзя. При выставлении неполного балла за задание в связи с наличием ошибок на усмотрение проверяющего предусмотрена возможность исправить работу на указанных в ответном письме условиях.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов (подробнее о плагиате см. на странице курса). Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

### Формат сдачи
Задания сдаются через систему Anytask. Инвайт можно получить у семинариста или ассистента. Присылать необходимо ноутбук с выполненным заданием. Сам ноутбук называйте в формате homework-practice-01-pandas-Username.ipynb, где Username — ваша фамилия.

Для удобства проверки самостоятельно посчитайте свою максимальную оценку (исходя из набора решенных задач) и укажите ниже.

Оценка: xx.

## 0. Введение

Сейчас мы находимся в jupyter-ноутбуке (или ipython-ноутбуке). Это удобная среда для написания кода, проведения экспериментов, изучения данных, построения визуализаций и других нужд, не связанных с написанием production-кода. 

Ноутбук состоит из ячеек, каждая из которых может быть либо ячейкой с кодом, либо ячейкой с текстом размеченным и неразмеченным. Текст поддерживает markdown-разметку и формулы в Latex.

Для работы с содержимым ячейки используется *режим редактирования* (*Edit mode*, включается нажатием клавиши **Enter** после выбора ячейки), а для навигации между ячейками искользуется *командный режим* (*Command mode*, включается нажатием клавиши **Esc**). Тип ячейки можно задать в командном режиме либо с помощью горячих клавиш (**y** to code, **m** to markdown, **r** to edit raw text), либо в меню *Cell -> Cell type*. 

После заполнения ячейки нужно нажать *Shift + Enter*, эта команда обработает содержимое ячейки: проинтерпретирует код или сверстает размеченный текст.

In [2]:
# ячейка с кодом, при выполнении которой появится output
2 + 2

4

Ячейка с неразмеченыным текстом.

Попробуйте создать свои ячейки, написать какой-нибудь код и текст какой-нибудь формулой.

In [ ]:
# your code

[Здесь](https://athena.brynmawr.edu/jupyter/hub/dblank/public/Jupyter%20Notebook%20Users%20Manual.ipynb) находится <s>не</s>большая заметка о используемом языке разметки Markdown. Он позволяет:

0. Составлять упорядоченные списки
1. Выделять *текст* <s>при</s> **необходимости**
2. Добавлять [ссылки](http://imgs.xkcd.com/comics/the_universal_label.png)


* Составлять неупорядоченные списки

Делать вставки с помощью LaTex:
    
$
\left\{
\begin{array}{ll}
x = 16 \sin^3 (t) \\ 
y = 13 \cos (t) - 5 \cos (2t) - 2 \cos (3t) - \cos (4t) \\
t \in [0, 2 \pi]
\end{array}
\right.$

А ещё можно вставлять картинки, или гифки, или что захотите:

<img src="https://media1.tenor.com/images/9b8fbe9214504bcf6c60fe4e4f7e114e/tenor.gif?itemid=5416416" style="width: 400px">

### Google Colab

Что за колаб? 

**Google Colab (Colaboratory)** это **Jupyter Notebook + Cloud + Google Drive.**

Компания Google предоставляет возможность бесплатно запускать ноутбуки (предварительно загрузив их на свой гугл-диск) прямо в облаке. При этом вам не требуется установка никаких пакетов на свою машину, а работать можно прямиком из браузера. Вот ссылка:

https://colab.research.google.com

При использовании вы увидете много сходств с jupyter ноутбуком. Одним из преимуществ является доступность GPU, соответствующую опцию можно активировать в настройках сервиса. При желании вы сможете найти в интернете много туториалов по использованию или разобраться самостоятельно =)

## 1. Табличные данные и Pandas

Pandas — удобная библиотека для работы с табличными данными в Python, если данных не слишком много и они помещаются в оперативную память вашего компьютера. Несмотря на неэффективность реализации и некоторые проблемы, библиотека стала стандартом в анализе данных.

Основной объект в pandas — это DataFrame, представляющий собой таблицу с именованными колонками различных типов, индексом (может быть многоуровневым). DataFrame можно создавать, считывая таблицу из файла или задавая вручную из других объектов.

В этой части потребуется выполнить несколько небольших заданий. Можно пойти двумя путями: сначала изучить материалы, а потом приступить к заданиям, или же разбираться "по ходу". Выбирайте сами.

Материалы:
1. [Pandas за 10 минут из официального руководства](http://pandas.pydata.org/pandas-docs/stable/10min.html)
2. [Документация](http://pandas.pydata.org/pandas-docs/stable/index.html) (стоит обращаться, если не понятно, как вызывать конкретный метод)
3. [Примеры использования функционала](http://nbviewer.jupyter.org/github/justmarkham/pandas-videos/blob/master/pandas.ipynb)

Многие из заданий можно выполнить несколькими способами. Не существуют единственно верного, но попробуйте максимально задействовать арсенал pandas и ориентируйтесь на простоту и понятность вашего кода. Мы не будем подсказывать, что нужно использовать для решения конкретной задачи, попробуйте находить необходимый функционал сами (название метода чаще всего очевидно). В помощь вам документация, поиск и stackoverflow.

In [629]:
%pylab inline
import pandas as pd

Populating the interactive namespace from numpy and matplotlib


Скачаем данные:

In [630]:
!wget  -O 'end_seminar.xlsx' -q 'https://www.dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=0'

"wget" ­Ґ пў«пҐвбп ў­гв७­Ґ© Ё«Ё ў­Ґи­Ґ©
Є®¬ ­¤®©, ЁбЇ®«­пҐ¬®© Їа®Ја ¬¬®© Ё«Ё Ї ЄҐв­л¬ д ©«®¬.


Для пользователей Windows: скачайте файл самостоятельно и поместите его в папку с тетрадкой. Или попробуйте один из следующих вариантов:

In [ ]:
# !powershell iwr -outf somefile https://somesite/somefile

In [ ]:
# !pip install wget
# import wget 
# wget.download('https://dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=1', 'end_seminar.xlsx')

##### В первой части задания (до раздела "Распределение студентов по курсам") использование циклов запрещается и повлечет за собой снижение оценки. Использование <code>vectorize</code> и  <code>apply</code>, <code>apply_along_axis</code> крайне нежелательно.

Для каждой задачи из этого раздела вы должны написать код для получения ответа, а также дать текстовый ответ, если он предполагается.

На некоторые вопросы вы можете получить путём пристального взгляда на таблицу, но это не будет засчитываться. Вы в любом случае должны получить ответ с помощью кода.

#### 1. [0 баллов] Откройте файл с таблицей (не забудьте про её формат). Выведите последние 10 строк.

Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк, какие столбцы? (на это не надо отвечать, просто подумайте об этом)

In [631]:
data = pd.read_excel('input/01_end_seminar.xlsx')
data.head()

timestamp                                id  rating  18_group  \
0 2020-05-15 01:12:50.543  93ff79a51cd602f1dd3028ba2c129503     698     181.0   
1 2020-05-15 02:46:48.066  26b01b1c4cd5656bab18d24c548834fb     647     181.0   
2 2020-05-15 03:12:41.480  30f3653fc176d54e89ac3179c455c6dd     624     185.0   
3 2020-05-15 04:43:08.994  1528f0eaa027580820ccf0d92a53ad68     579     182.0   
4 2020-05-15 07:47:17.197  496ea4f0d4abe264b1bb1b80eb3830c5     572     183.0   

   is_mi                             fall_1  \
0    NaN                           Язык SQL   
1    NaN  Высокопроизводительные вычисления   
2    NaN   Безопасность компьютерных систем   
3    NaN        Statistical Learning Theory   
4    NaN  Высокопроизводительные вычисления   

                              fall_2                         fall_3  \
0  Высокопроизводительные вычисления           Матричные вычисления   
1   Безопасность компьютерных систем                       Язык SQL   
2               Матричные вычисления  Моделирование временных рядов   
3  Высокопроизводительные вычисления           Матричные вычисления   
4   Безопасность компьютерных систем              Теория баз данных   

                 spring_1                spring_2             spring_3  \
0  Дискретная оптимизация        Численные методы  Машинное обучение 2   
1  Дискретная оптимизация        Численные методы  Машинное обучение 2   
2  Дискретная оптимизация     Машинное обучение 2     Численные методы   
3  Дискретная оптимизация        Численные методы  Машинное обучение 2   
4       Компьютерные сети  Дискретная оптимизация     Численные методы   

  is_first_time  percentile  17_group blended  is_ml_student  
0            Да    0.183857       NaN     NaN            NaN  
1           Нет    0.426009       NaN     NaN            NaN  
2            Да    0.538117       NaN     NaN            NaN  
3            Да    0.744395       NaN     NaN            NaN  
4            Да    0.780269       NaN     NaN            NaN

#### 2. [0.5 балла] Ответьте на вопросы:
1. Сколько было заявок из групп 18-го года набора, а сколько из групп 17-го года?

In [632]:
print(f'Количество заявок 18-го года: {data["18_group"].notna().sum()}')
print(f'Количество заявок 17-го года: {data["17_group"].notna().sum()}')

Количество заявок 18-го года: 223
Количество заявок 17-го года: 138


2. Есть ли студенты с равными перцентилями?

In [633]:
len(data['percentile']) != len(data['percentile'].unique())

True

#### 3. [0.5 балла] Есть ли в данных пропуски? В каких колонках? Сколько их в каждой из этих колонок?

In [634]:
data.isna().sum()

timestamp          0
id                 0
rating             0
18_group         138
is_mi            343
fall_1             0
fall_2             0
fall_3             0
spring_1           0
spring_2           0
spring_3           0
is_first_time      2
percentile         0
17_group         223
blended          223
is_ml_student    304
dtype: int64

Заполните пропуски пустой строкой для строковых колонок и нулём для числовых.

In [635]:
data[list(data.select_dtypes(exclude='object'))] = data.select_dtypes(exclude='object').fillna(value=0)
data[list(data.select_dtypes(include='object').columns)] = data.select_dtypes(include='object').fillna(value='')

In [636]:
data.isnull().sum()

timestamp        0
id               0
rating           0
18_group         0
is_mi            0
fall_1           0
fall_2           0
fall_3           0
spring_1         0
spring_2         0
spring_3         0
is_first_time    0
percentile       0
17_group         0
blended          0
is_ml_student    0
dtype: int64

#### 4. [0.5 балла] Посмотрите повнимательнее на колонку 'is_first_time'. 

Есть ли в ней ответы "Нет"? Сколько их?

In [637]:
data['is_first_time'].value_counts()

Да     308
Нет     51
         2
Name: is_first_time, dtype: int64

Если вы найдете повторные обращения студентов, оставьте только самую позднюю версию. <i>Обращения со значением "Нет" в <code>is_first_time</code> могут быть как повторными, так и первичными, поскольку поле заполняли сами студенты.</i>

In [638]:
data.loc[data['id'].isin(data[data['id'].duplicated()]['id'])]

timestamp                                id  rating  \
22  2020-05-15 22:15:31.902  a65d50ef4fadfac64508a54e972d8623     626   
23  2020-05-15 22:30:31.902  a65d50ef4fadfac64508a54e972d8623     626   
29  2020-05-16 14:46:06.274  b121e4d7d62387cd5199cca464a34a6d     694   
30  2020-05-16 15:01:06.274  b121e4d7d62387cd5199cca464a34a6d     694   
97  2020-05-18 18:38:41.094  5ee2ad618729a6dc3bdb7f2858826ef0     628   
98  2020-05-18 18:53:41.094  5ee2ad618729a6dc3bdb7f2858826ef0     628   
121 2020-05-18 20:47:28.759  5c49473012411a1eee72f8ea63767adf     573   
126 2020-05-18 21:02:28.759  5c49473012411a1eee72f8ea63767adf     573   
129 2020-05-18 21:12:02.445  ffc58d371e2f52e1c7f45d49d3d629d0     659   
139 2020-05-18 21:27:02.445  ffc58d371e2f52e1c7f45d49d3d629d0     659   
141 2020-05-18 21:34:22.233  f8fab519d78047712c06798c8b7fe38f     561   
144 2020-05-18 21:36:52.845  9015b809a29fb65ce23d7addf16d770d     665   
150 2020-05-18 21:49:22.233  f8fab519d78047712c06798c8b7fe38f     561   
152 2020-05-18 21:51:52.845  9015b809a29fb65ce23d7addf16d770d     665   
174 2020-05-18 22:36:34.323  8aa0f1b7331b49464174b02cd693be52     567   
176 2020-05-18 22:51:34.323  8aa0f1b7331b49464174b02cd693be52     567   
186 2020-05-18 23:20:28.235  86c24daaa62246142d46ee58534a65e0     530   
192 2020-05-18 23:35:28.235  86c24daaa62246142d46ee58534a65e0     530   
196 2020-05-18 23:40:41.196  47e2d8bbb630bc2d9b3fbf98fe77cffb     616   
204 2020-05-18 23:55:41.196  47e2d8bbb630bc2d9b3fbf98fe77cffb     616   
257 2020-05-17 15:30:40.828  906692ccd33f59058380aac1598c5480    1143   
259 2020-05-17 15:55:40.828  906692ccd33f59058380aac1598c5480    1143   
300 2020-05-18 19:58:36.111  9a66d84568b8017d98eb2e678b57296b    1205   
302 2020-05-18 20:23:36.111  9a66d84568b8017d98eb2e678b57296b    1205   
321 2020-05-18 22:15:28.531  d9c068e5578253c2b0721558dd0813e1    1075   
325 2020-05-18 22:40:28.531  d9c068e5578253c2b0721558dd0813e1    1075   
333 2020-05-18 22:59:48.715  50f6d5b6a7fd6c8b1b8959f9ce1f808d    1086   
342 2020-05-18 23:24:48.715  50f6d5b6a7fd6c8b1b8959f9ce1f808d    1086   

     18_group  is_mi                                           fall_1  \
22      184.0    0.0                Высокопроизводительные вычисления   
23      184.0    0.0                 Безопасность компьютерных систем   
29      185.0    0.0                                Теория баз данных   
30      185.0    0.0                 Безопасность компьютерных систем   
97      185.0    0.0                                         Язык SQL   
98      185.0    0.0                 Безопасность компьютерных систем   
121     183.0    0.0                                         Язык SQL   
126     183.0    0.0                 Безопасность компьютерных систем   
129     187.0    0.0  Сбор и обработка данных с помощью краудсорсинга   
139     187.0    0.0                             Матричные вычисления   
141     185.0    1.0                                Теория баз данных   
144     188.0    1.0                    Моделирование временных рядов   
150     185.0    1.0                    Моделирование временных рядов   
152     188.0    1.0                                         Язык SQL   
174     187.0    0.0                                         Язык SQL   
176     187.0    0.0                 Безопасность компьютерных систем   
186     184.0    0.0                    Моделирование временных рядов   
192     184.0    0.0  Сбор и обработка данных с помощью краудсорсинга   
196     188.0    0.0                                         Язык SQL   
204     188.0    0.0  Сбор и обработка данных с помощью краудсорсинга   
257       0.0    0.0            Байесовские методы машинного обучения   
259       0.0    0.0                Анализ неструктурированных данных   
300       0.0    0.0             Глубинное обучение в обработке звука   
302       0.0    0.0                Анализ неструктурированных данных   
321       0.0    0.0                      Statistical Learning Theory   
32

In [639]:
# Индексы которые должны остаться
ind = data[data['id'].duplicated()].index
ind

Int64Index([23, 30, 98, 126, 139, 150, 152, 176, 192, 204, 259, 302, 325, 342], dtype='int64')

In [640]:
data.sort_values(by=['timestamp'], inplace=True)
data.drop_duplicates(subset=['id'], keep = 'last', inplace = True) #так как мы отсортировали по времени наши строки, то с помощью аргумента keep, мы оставим только последнюю из всех версий по одинаковому id.

#### 5. [0.5 балла] Какие  blended-курсы для четверокурсников существуют? На какой blended-курс записалось наибольшее количество студентов? На каком из курсов собрались студенты с самым высоким средним рейтингом? 

In [641]:
data.query('`18_group` == 0').blended.value_counts()

DevOps                                   87
Соревновательный анализ данных           32
Введение в дифференциальную геометрию    15
Name: blended, dtype: int64

In [642]:
data.query('`18_group` == 0').groupby(by='blended').mean()['rating']

blended
DevOps                                   1154.172414
Введение в дифференциальную геометрию    1210.733333
Соревновательный анализ данных           1144.218750
Name: rating, dtype: float64

#### 6. [1 балл] Выясните, есть ли в данных студенты с абсолютно одинаковыми предпочтениями по всем курсам (не забудьте учесть blended-курсы для четверокурсников). Сколько таких наборов, которые взяли несколько студентов? Выведите их все вместе с количеством выбравших их студентов.

<i>Предпочтения двух студентов считаются абсолютно одинаковыми, если выбранные ими дисциплины имеют одинаковый приоритет.</i>

In [643]:
data_dublicates = data.duplicated(subset=['fall_1', 'fall_2',
       'fall_3', 'spring_1', 'spring_2', 'spring_3', 'blended'], keep=False)

len(data[data_dublicates])

32

In [644]:
data[data_dublicates].groupby(['fall_1', 'fall_2','fall_3', 'spring_1', 'spring_2', 'spring_3', 'blended'], as_index=False).size()

fall_1  \
0                       Statistical Learning Theory   
1             Байесовские методы машинного обучения   
2             Байесовские методы машинного обучения   
3                  Безопасность компьютерных систем   
4                  Безопасность компьютерных систем   
5                  Безопасность компьютерных систем   
6                  Безопасность компьютерных систем   
7                  Безопасность компьютерных систем   
8                  Безопасность компьютерных систем   
9                  Безопасность компьютерных систем   
10                Высокопроизводительные вычисления   
11                Высокопроизводительные вычисления   
12                    Моделирование временных рядов   
13  Сбор и обработка данных с помощью краудсорсинга   
14                                         Язык SQL   

                                             fall_2  \
0                 Анализ неструктурированных данных   
1                 Анализ неструктурированных данных   
2                 Анализ неструктурированных данных   
3                       Statistical Learning Theory   
4                 Высокопроизводительные вычисления   
5                 Высокопроизводительные вычисления   
6                 Высокопроизводительные вычисления   
7                              Матричные вычисления   
8   Сбор и обработка данных с помощью краудсорсинга   
9                                 Теория баз данных   
10                 Безопасность компьютерных систем   
11  Сбор и обработка данных с помощью краудсорсинга   
12  Сбор и обработка данных с помощью краудсорсинга   
13                                         Язык SQL   
14                Высокопроизводительные вычисления   

                                             fall_3  \
0   Сбор и обработка данных с помощью краудсорсинга   
1              Глубинное обучение в обработке звука   
2   Сбор и обработка данных с помощью краудсорсинга   
3                              Матричные вычисления   
4                 Анализ неструктурированных данных   
5                                          Язык SQL   
6                                          Язык SQL   
7                     Моделирование временных рядов   
8                              Матричные вычисления   
9                                          Язык SQL   
10  Сбор и обработка данных с помощью краудсорсинга   
11                    Моделирование временных рядов   
12                 Безопасность компьютерных систем   
13                                Теория баз данных   
14                                Теория баз данных   

                                             spring_1  \
0                            Обучение с подкреплением   
1                            Обучение с подкреплением   
2                            Обучение с подкреплением   
3                              Дискретная оптимизация   
4   Проектирование и разработка высоконагруженных ...   
5                                   Компьютерные сети   
6         Промышленное программирование на языке Java   
7         Промышленное программирование на языке Java   
8                                   Компьютерные сети   
9                                 Машинное обучение 2   
10  Проектирование и разработка высоконагруженных ...   
11  Проектирование и разработка высоконагруженных ...   
12                                  Компьютерные сети   
13        Промышленное программирование на языке Java   
14                                  Компьютерные сети   

                                       spring_2  \
0                        Конфликты и кооперация   
1       Генеративные модели в машинном обучении   
2                           Компьютерное зрение   
3   Промышленное программирование на языке Java   
4          Методы сжатия и передачи медиаданных   
5   Промышленное программирование на языке Java   
6                             Компьютерные сети   
7                             Компьютерные сети   
8   Промышленное программирова

In [645]:
len(data[data_dublicates].groupby(['fall_1', 'fall_2','fall_3', 'spring_1', 'spring_2', 'spring_3', 'blended'],as_index=False).size())

15

#### 7. [0.5 балла] Найдите курсы по выбору, на которые записывались как студенты 18-го года набора, так и студенты 17-го года.

In [ ]:
# your code

Методом исключения найдите курсы, которые предлагались только студентам 18-го года и только студентам 17-го года.

In [ ]:
# your code

### Визуализации и matplotlib

При работе с данными часто неудобно делать какие-то выводы, если смотреть на таблицу и числа в частности, поэтому важно уметь визуализировать данные. Здесь будут описаны ключевые правила оформления графиков для **всех** домашних заданий.

У matplotlib, конечно же, есть [документация](https://matplotlib.org/users/index.html) с большим количеством [примеров](https://matplotlib.org/examples/), но для начала достаточно знать про несколько основных типов графиков:
- plot — обычный поточечный график, которым можно изображать кривые или отдельные точки;
- hist — гистограмма, показывающая распределение некоторой величины;
- scatter — график, показывающий взаимосвязь двух величин;
- bar — столбцовый график, показывающий взаимосвязь количественной величины от категориальной.

В этом задании вы попробуете построить один из них. Не забывайте про базовые принципы построения приличных графиков:
- оси должны быть подписаны, причём не слишком мелко;
- у графика должно быть название;
- если изображено несколько графиков, то необходима поясняющая легенда;
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах);
- сетка на графике помогает оценить значения в точках на глаз, это обычно полезно, поэтому лучше ее отрисовывать;
- если распределение на гистограмме имеет тяжёлые хвосты, лучше использовать логарифмическую шкалу.

Полезные строчки:
- для красивой картинки <code>%config InlineBackend.figure_format = 'retina'</code>
- задаем размер (почти) всех графиков (можно выбрать другие значения) <code>plt.rcParams['figure.figsize'] = 8, 5</code>
- размер шрифта подписей графиков <code>plt.rcParams['font.size'] = 12</code>
- формат в котором сохраняется изображение <code>mpl.rcParams['savefig.format'] = 'pdf'</code>
- sns – seaborn, добавляет решетку <code>sns.set_style('darkgrid')</code>

In [646]:
%matplotlib inline
import matplotlib.pyplot as plt

#### 8. [0.5 балла] Постройте график средних рейтингов по дням получения ответов (bar plot).

In [628]:
# your code

Сохраните график в формате pdf (так он останется векторизованным).

In [ ]:
# your code

Еще одна билиотека для визуализации: [seaborn](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html). Это настройка над matplotlib, иногда удобнее и красивее делать визуализации через неё. 

### 2. Распределение студентов по курсам.

Теперь вам нужно распределить студентов по осенним курсам по выбору, учитывая их предпочтения.

Алгоритм распределения студентов по курсам:
1. По умолчанию на каждой дисциплине по выбору у 3 и 4 курсов может учиться 1 группа (до 30 студентов). Исключения описаны ниже. На blended-дисциплинах для четверокурсников количество мест не ограничено.
2. Проводится первая волна отбора. Для каждой дисциплины формируется список тех, кто указал её первым приоритетом (если студент должен выбрать два курса по выбору, то для него дисциплины, которые он указал первым и вторым приоритетом, рассматриваются как дисциплины первого приоритета). Если желающих больше, чем мест, то выбирается топ по перцентилю рейтинга.
3. На дисциплинах, где остались места после первой волны, формируются списки тех, кто выбрал их вторым приоритетом, и места заполняются лучшими по перцентили рейтинга студентами. После этого проводится такая же процедура для дисциплин третьего приоритета.
4. Если студент не попал на необходимое количество курсов по итогам трёх волн, с ним связывается учебный офис и решает вопрос в индивидуальном порядке.

Обращаем ваше внимание на следующие детали:

- По умолчанию студент выбирает один осенний и один весенний курс по выбору, а также четверокурсники выбирают один blended-курс. Студенты групп 17-го года специализаций МОП и ТИ выбирают по 2 осенних и 2 весенних курса по выбору, также студенты групп 18' специализации МИ выбирают 2 осенних курса. <i>Для студентов, которые выбирают 2 курса (например, осенних) первый приоритет — <code>fall_1</code> и <code>fall_2</code>, второй приоритет — <code>fall_3</code>. Такие студенты участвуют только в двух волнах отбора</i>.

- Студенты специализации МОП не могут выбрать весенним курсом по выбору Машинное обучение 2. <i>Если студент специализации МОП выбрал Машинное обучение 2, то его приоритеты сдвигаются. Из-за совпадений первого и второго курса по выбору двигать приоритеты не надо</i>.

- Blended-курсы не трогайте, по ним не надо распределять, на другие курсы они никак не влияют.

- Постарайтесь воздержаться от использования циклов там, где это возможно. <i>Допустимо итерироваться по <b>курсам</b>, на которые проводится отбор, и по <b>волнам</b> отбора. Если вы придумаете, как обойтись и без этих циклов, то на усмотрение проверяющего могут быть добавлены бонусные баллы. <b>Дублирование кода не признается успешным избавлением от циклов</b></i>

- На выходе ожидается файл res_fall.csv с результатами распределения на осенние курсы по выбору. Файл должен быть следующего формата:

    * три столбца: ID, course1, course2
    
    * Если студент не попал на курс, но должен был, то вместо названия курса в ячейке должна быть строка "???"
    
    * Если студент должен выбрать только один курс, то в колонке course2 для него должна стоять строка "-"
    
    * Если студент должен выбрать два курса по выбору, то порядок в колонках course1 и course2 не важен.
    
    * Формат csv: для сохранения воспользуйтесь df.to_csv('solution.csv', index=None)
    

Для работы вам могут понадобиться следующие данные:

- Результаты опроса (вы уже использовали этот файл в первой части задания, но на всякий случай ссылка: https://www.dropbox.com/s/f4rm8sjc3v99p0m/_end_seminar.xlsx?dl=0)

- Соответствие номеров групп специализациям:

    * 171, 172 - МОП; 173 - ТИ; 174 — АДИС; 175, 176 — РС; 177 — АПР
    
    * У студентов 18-го года номера групп соответствуют номерам до распределения по специализациям.

- Ограничения по количеству мест на курсах по выбору:

    * Осенние: везде 30 мест, кроме Statistical Learning Theory (60 мест), Высокопроизводительных вычислений (60 мест), Анализа неструктурированных данных ($\infty$ мест)

    * Весенние: везде 30 мест, кроме Обучения с подкреплением (60 мест), Анализа данных в бизнесе (60 мест).


Кстати, убедитесь, что в данных больше нет пропусков и повторных записей.

#### 0. Проверка

Для начала давайте убедимся, что вы успешно выполнили задания первой части и проверим ваши данные на наличие пропусков и повторов:

In [647]:
assert data.shape[0] == 347, 'В таблице остались повторы или потеряны данные'

assert data.isna().sum().sum() == 0, 'В таблице остались пропуски'

Если вы не получили AssertionError, то можете продолжать.

#### 1. [1 балл] Создайте новый признак, обозначающий, сколько осенних курсов должен выбрать студент

В этом вам может помочь информация о специализации и группе стундента.

In [649]:
data['fall_courses'] = np.ones(shape=(data.shape[0],1), dtype=numpy.int8)

In [650]:
data[(data['18_group'] == 0) & (data['is_mi'] == 1)].shape

(0, 17)

Проверка:

In [ ]:
col_name =     # insert your new column name as str

assert(df[df['id'] == '2662600c2c37e11e62f6ee0b88452f22'][col_name] == 2).all()
assert(df[df['id'] == 'd555d2805e1d93d4f023e57dc4c8f403'][col_name] == 2).all()
assert(df[df['id'] == '8fe79f84f36e3a5d2d6745621321302c'][col_name] == 1).all()
assert(df[df['id'] == 'e4caca755ee0bdd711e18fb8084958b5'][col_name] == 1).all()

#### 2. [2 балла] Распределите студентов в соответствии с первым приоритетом

In [ ]:
# (￣^￣)ゞ

Здесь для проверки приведена таблица, в которой есть 2 дополнительные колонки:
    
    1) is_first_place - является ли студент лучшим по перцентили хотя бы на одном из курсов, куда он был зачислен 
    (True / NaN)
    
    2) is_last_place  - является ли студент худшим по перцентили хотя бы на одном из курсов, куда он был зачислен (True / NaN)

In [ ]:
!wget  -O '2_task_check.csv' -q 'https://www.dropbox.com/s/v8o2zzq3iz5gc9w/_2_task_check.csv?dl=0'
check_df = pd.read_csv('2_task_check.csv')

После распределения студентов в соответствии с первым приоритетом добавьте в свой датафрейм аналогичные признаки и запустите проверку:

In [ ]:
fir_col_name =       # insert name of your new column with is_first_place as str
last_col_name =      # insert name of your new column with is_last_place as str


assert((df[df[fir_col_name].isna() == False][['id']].sort_values('id').reset_index(drop=True)
        ==
        check_df[check_df['is_first_place'].isna() == False][['id']].sort_values('id').reset_index(drop=True)
       ).id.values).all()


assert((df[df[last_col_name].isna() == False][['id']].sort_values('id').reset_index(drop=True)
       == 
       check_df[check_df['is_last_place'].isna() == False][['id']].sort_values('id').reset_index(drop=True)
      ).id.values).all()

#### 3. [3 балла] Проведите все три волны отбора студентов на курсы по выбору

In [ ]:
# (￣^￣)ゞ

**Отправьте свой файл res_fall.csv в контест (https://contest.yandex.ru/contest/29176/problems/A/) и прикрепите/укажите ниже ваш никнейм и ссылку на успешную посылку.**

*Дисклеймер:*

Контест выдаётся для самопроверки. Если ваша посылка получила ОК, то код, скорее всего, правильный. Но при этом оценка всё равно может быть снижена в случае обнаружения неэффективностей или ошибок в коде. Если вы сдадите в AnyTask очевидно неработающий код или ноутбук без кода, но при этом в контест будет сдан корректный файл, то это будет расцениваться как плагиат.

На всякий случай просим вас сдать вместе с ноутбуком файл res_fall.csv в anytask

**Дополнительное задание. [2 бонусных балла] Распределите таким же образом студентов еще и на весенние курсы по выбору.**

Если ваш код был хорошо структурирован, то это не составит проблем. 

Если вы выполнили это задание, сдайте среди прочего файл res_spring.csv в таком же формате, как и res_fall.csv.

In [ ]:
# (￣^'￣)ゞ

**Отправьте свой файл res_spring.csv в контест (https://contest.yandex.ru/contest/29176/problems/B/) и прикрепите/укажите ниже ваш никнейм и ссылку на успешную посылку.**

На всякий случай просим вас сдать вместе с ноутбуком файл res_spring.csv в anytask

Вставьте картинку, описывающую ваш опыт выполнения этого задания: